Improve NSScreen screens error handling across codebase

zakariaridouh opened the pull request:
#### d769e127f9e9048e401b0980d210e1e5f09690a2
<pre>
Improve NSScreen screens error handling across codebase
<a href="https://bugs.webkit.org/show_bug.cgi?id=291147">https://bugs.webkit.org/show_bug.cgi?id=291147</a>
<a href="https://rdar.apple.com/148673358">rdar://148673358</a>

Reviewed by NOBODY (OOPS!).

After bug 291122, I noticed theres a few inconsistencies
in how we query for screens from the AppKit `NSScreen screens` API,
and that in some cases we are not validating if we have a screen.

Make all of the calls consistent, and add some debug asserts
to validate that we correctly got screens, and release asserts
in some testing code.

i.e, use `[[NSScreen screens] firstObject]` as it
returns `nil` if no screens are given.

* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::firstScreen):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(getFirstScreen):
(-[WKFullScreenWindowController _continueEnteringFullscreenAfterPostingNotification:]):
(-[WKFullScreenWindowController _continueExitingFullscreenAfterPostingNotificationAndExitImmediately:]):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(getFirstScreen):
(-[WebFullScreenController applicationDidResignActive:]):
(-[WebFullScreenController enterFullScreen:willEnterFullscreen:didEnterFullscreen:]):
(-[WebFullScreenController _updateMenuAndDockForFullScreen]):
* Source/WebKitLegacy/mac/WebView/WebVideoFullscreenController.mm:
(getFirstScreen):
(-[WebVideoFullscreenController applicationDidResignActive:]):
* Source/WebKitLegacy/mac/WebView/WebWindowAnimation.mm:
(getFirstScreen):
(flipRect):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(getFirstScreen):
(createWebViewAndOffscreenWindow):
* Tools/DumpRenderTree/mac/EventSendingController.mm:
(getFirstScreen):
(-[EventSendingController mouseScrollByX:andY:continuously:]):
(-[EventSendingController sendScrollEventAt:deltaX:deltaY:units:wheelPhase:momentumPhase:timestamp:]):
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(coreGraphicsScreenRectForAppKitScreenRect):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(getFirstScreen):
(-[TestWKWebView wheelEventAtPoint:wheelDelta:]):
* Tools/TestWebKitAPI/mac/OffscreenWindow.mm:
(getFirstScreen):
(-[OffscreenWindow initWithSize:isKeyWindow:]):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(WTR::getFirstScreen):
(WTR::EventSenderProxy::mouseScrollBy):
(WTR::EventSenderProxy::mouseScrollByWithWheelAndMomentumPhases):
(WTR::EventSenderProxy::sendWheelEvent):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d769e127f9e9048e401b0980d210e1e5f09690a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/146313 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/18990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/12334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/154980 "Built successfully") | [💥 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/99760 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/4dcf27b6-be00-49f4-963a-9d8f3ea89bb2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/148188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/19463 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/18885 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/112597 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/99760 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/44098ad1-73ba-45ac-9212-4eab2b0fc2fd) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/149276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/14952 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/131454 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/93466 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/11183f43-d5e1-41fd-811b-b9653badc237) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/14219 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/11973 "Found 1 new API test failure: TestWebKitAPI.ScrollbarTests.ScrollbarAvoidanceInConcentricContainerWithNonUniformCornerRadii (failure)") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/2426 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/123787 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/9193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/157301 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/472 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/10625 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/120626 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/18807 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/15758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/120924 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/18827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/130929 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/74593 "Built successfully") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/16598 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/7971 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/18428 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/82180 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/18158 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/18323 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/18216 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
```diff
@@ -239,8 +239,10 @@ - (void)enterFullScreen:(CompletionHandler<void(bool)>&&)completionHandler
     NSRect webViewFrame = convertRectToScreen([_webView window], [_webView convertRect:[_webView frame] toView:nil]);
 
+    NSScreen *firstScreen = [[NSScreen screens] firstObject];
+    ASSERT_WITH_MESSAGE(firstScreen, "No screens found, possibly due to no WindowServer session. This configuration is not supported.");
```
Looks good to me overall, but I'm not sure about the point of debug asserts. We are not going to hit this when debugging at desk, it could be a strange situation that happens in the wild. So we need an audit trail that's visible in a sysdiagnose, and isn't compiled out.

zakariaridouh replied:
Should we add `RELEASE_ASSERT_WITH_MESSAGE` to all these sites? I'm a little bit concerned about performance in non test parts of the code, so maybe `RELEASE_ASSERT`s in the testing code, and nothing elsewhere?

Otherwise, we could do `RELEASE_ASSERT`s everywhere, with an A/B test to make sure no performance regressions.

aproskuryakov replied:
I was thinking `RELEASE_LOG`, but I'll leave this for others to weigh on it. Seems unlikely to be very performance sensitive though, we wouldn't be querying screens in a loop.

ddkilzer replied:
My recommendation is to use `RELEASE_ASSERT_WITH_MESSAGE()` because Debug asserts are essentially unmaintained at this point.

ddkilzer commented on Source/WebCore/platform/mac/PlatformScreenMac.mm:
```diff
@@ -78,10 +78,7 @@ static PlatformDisplayID displayID(Widget* widget)
-    if (![screens count])
-        return nil;
-    return [screens objectAtIndex:0];
+    return [[NSScreen screens] firstObject];
```
Why assert elsewhere but not here?  Is it because callers of `firstScreen()` handle a `nil` result?

zakariaridouh replied:
I was originally slightly hesitant to use release asserts in this non testing code, but looked at the calling sites and I can see we're not in hot code nor is this codepath being called in a loop. We should release assert here as well.

ddkilzer reviewed (approved):
Overall this looks fine, but I'm curious why we only add assertions in some places but not others.

I think we should probably switch to release assertions consistently.  Also, there are enough code sites that it would be best to extract this into a helper function (even if it's one helper function per-project) so you don't have to copy/paste the same code everywhere.

zakariaridouh committed:
Improve NSScreen screens error handling across codebase
https://bugs.webkit.org/show_bug.cgi?id=291147
rdar://148673358

Reviewed by NOBODY (OOPS!).

After bug 291122, I noticed theres a few inconsistencies
in how we query for screens from the AppKit `NSScreen screens` API,
and that in some cases we are not validating if we have a screen.

Make all of the calls consistent, and add some debug asserts
to validate that we correctly got screens, and release asserts
in some testing code.

i.e, use `[[NSScreen screens] firstObject]` as it
returns `nil` if no screens are given.

* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::firstScreen):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(getFirstScreen):
(-[WKFullScreenWindowController _continueEnteringFullscreenAfterPostingNotification:]):
(-[WKFullScreenWindowController _continueExitingFullscreenAfterPostingNotificationAndExitImmediately:]):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(getFirstScreen):
(-[WebFullScreenController applicationDidResignActive:]):
(-[WebFullScreenController enterFullScreen:willEnterFullscreen:didEnterFullscreen:]):
(-[WebFullScreenController _updateMenuAndDockForFullScreen]):
* Source/WebKitLegacy/mac/WebView/WebVideoFullscreenController.mm:
(getFirstScreen):
(-[WebVideoFullscreenController applicationDidResignActive:]):
* Source/WebKitLegacy/mac/WebView/WebWindowAnimation.mm:
(getFirstScreen):
(flipRect):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(getFirstScreen):
(createWebViewAndOffscreenWindow):
* Tools/DumpRenderTree/mac/EventSendingController.mm:
(getFirstScreen):
(-[EventSendingController mouseScrollByX:andY:continuously:]):
(-[EventSendingController sendScrollEventAt:deltaX:deltaY:units:wheelPhase:momentumPhase:timestamp:]):
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(coreGraphicsScreenRectForAppKitScreenRect):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(getFirstScreen):
(-[TestWKWebView wheelEventAtPoint:wheelDelta:]):
* Tools/TestWebKitAPI/mac/OffscreenWindow.mm:
(getFirstScreen):
(-[OffscreenWindow initWithSize:isKeyWindow:]):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(WTR::getFirstScreen):
(WTR::EventSenderProxy::mouseScrollBy):
(WTR::EventSenderProxy::mouseScrollByWithWheelAndMomentumPhases):
(WTR::EventSenderProxy::sendWheelEvent):
